Database Package Refactoring (Clean Version)

orware opened the pull request:
Significant refactoring of the database package. Reorganized the various types of objects into folders (driver, query, exporter, importer). Added in a new base PDO class that can be extended from rather easily to create new drivers. As an example, a new Oracle PDO-based driver has been created. A new Preparable Interface for Queries has been added, allowing you to bind variables to a query. A new Limitable Interface for Queries has been added, allowing the SQL query to have the special (usually Database Vendor specific) limiting method added to the SQL string.

orware committed:
Committing significant refactoring of the database package. Reorganized the various types of objects into folders (driver, query, exporter, importer). Added in a new base PDO class that can be extended from rather easily to create new drivers. As an example, a new Oracle PDO-based driver has been created. A new Preparable Interface for Queries has been added, allowing you to bind variables to a query. A new Limitable Interface for Queries has been added, allowing the SQL query to have the special (usually Database Vendor specific) limiting method added to the SQL string.

Conflicts:

	libraries/joomla/database/database/mysql.php
	libraries/joomla/database/database/mysqli.php
	libraries/joomla/database/database/mysqlimporter.php

orware committed:
Deleting all index.html files.

orware committed:
Committing the mysqli->close() fix provided by Elin in this pull request: https://github.com/joomla/joomla-platform/issues/570

orware committed:
Added in the truncateTable() method Michael added in this commit:
https://github.com/orware/joomla-platform/commit/67ada5c718beb7df7d6b6149c1b9c7651303d589

orware committed:
Removed 2 jimport() statements that were no longer needed from Rouven's commit here:
https://github.com/orware/joomla-platform/commit/8139bc01182a9f9bc73401b45b9d5a1c1a23c34e

orware committed:
Brought over Andrew's "Fix up missing scope declaration" changes from this commit:
https://github.com/orware/joomla-platform/commit/f7db97534785428df9e8af7c8fc7264e180473b1#libraries/joomla/database

orware committed:
Added Michael's "Mark elements as JDatabaseQueryElement" changes from this commit:
https://github.com/orware/joomla-platform/commit/71bfd26f675dcf7b483fcdae5688a720c13d4e98#libraries/joomla/database

orware committed:
Added in Michael's "Proper return types for DB objects and getDbo" changes from this commit:
https://github.com/orware/joomla-platform/commit/71bfd26f675dcf7b483fcdae5688a720c13d4e98#libraries/joomla/database

orware committed:
Added Michael's "Add JOIN statement to UPDATE query" change from this commit:
https://github.com/orware/joomla-platform/commit/196d70f181c6cabdf585f703ba581f2b85e963a0#libraries/joomla/database

joomla-jenkins commented:
Test log missing. Tests failed to execute.
Checkstyle analysis reported 255 warnings and 117 errors.

orware committed:
Added in Elin's "Add missing __clone magic method to JDatabaseQuery and JDatabaseQueryElement." additions to the query class in these commits:
https://github.com/orware/joomla-platform/commit/3202485ac63de7fb82fdd87cd91bf423bacd719d
(I copied the methods from the latest version...I noticed two commits later you had an update to the method).

orware committed:
Added Nikolai's "Docblock fix" for JDatabase::getQuery() from this commit:
https://github.com/orware/joomla-platform/commit/ff29b837bf1421a01447756a675fc37b868b66cd#libraries/joomla/database

orware committed:
Changed hasUTF() @since tag back to 11.1

elkuku commented:
In case you feel lost about the 117 errors...
Here is some documentation: http://docs.joomla.org/Joomla_CodeSniffer
Please pay special attention to the note in usage related to the platform that is different to the (nonexistent) code style of the CMS  ;)
I have set up a **personal** page that might also help to resolve the checkstyle errors. About the unit tests - I am not sure as results seems to differ across different operating systems.
http://elkuku.github.com/pulltester/pulls/610 <= for reference only !

orware commented:
Thanks Nikolai, I've had trouble getting CodeSniffer to work and I haven't gone back to try again. If there's a quick way on somebody else's end to create the report for me I can correct the style errors and as many of the warnings as I can.

joomla-jenkins commented:
Build triggered by changes to the head.

Test log missing. Tests failed to execute.
Checkstyle analysis reported 255 warnings and 117 errors.

chdemko commented:
I would prefer to have a directory structure like this:

```
libraries/database/engine/mysql/server.php (JDatabaseEngineMysqlServer class)
                               /query.php (JDatabaseEngineMysqlQuery class)
                               /exporter.php (JDatabaseEngineMysqlExporter class)
                               /importer.php (JDatabaseEngineMysqlImporter class)
libraries/database/engine/postgres/server.php (JDatabaseEnginePostgresServer class)
                                  /query.php (JDatabaseEnginePostgresQuery class)
                                  /exporter.php (JDatabaseEnginePostgresExporter class)
                                  /importer.php (JDatabaseEnginePostgresImporter class)
```

This will allow the Joomla! autoloader to work without jimport and this will be easier for maintaining db engines (one folder for one engine).

orware commented:
The way the folders have been currently organized (and the files renamed),
aside from the ones in the table directory, since I didn't touch those,
should all conform to the new autoloader standard.

I don't agree to changing the word "driver" to "engine" as the driver
keyword more correctly signifies the purpose of those files and is the more
commonly used term.

Additionally, I know we have two different viewpoints on how to organize
the files into folders...the way that I've organized them, and then one
similar to the one you proposed.

I chose to organize them the way I did because it seemed more logical to me
(I didn't really need to change the class names at all, only the filenames
once I placed them in the correct folder), considering the types that I saw
(driver, query, importer, exporter type files) and because I liked the way
the driver type was at the very end of the class name instead of in the
middle, but that's just my preference.

If others feel strongly about one organization over the other it might be
good for them to chime to see if there's a strong argument for one over the
other (I simply wanted to organize things according to the new autoloader
without having to modify class names).

On Tue, Dec 6, 2011 at 6:40 AM, Christophe Demko <
reply@reply.github.com

> wrote:
> 
> I would prefer to have a directory structure like this:
> 
>    libraries/database/engine/mysql/server.php (JDatabaseEngineMysqlServer
> class)
>                                   /query.php (JDatabaseEngineMysqlQuery
> class)
>                                   /exporter.php
> (JDatabaseEngineMysqlExporter class)
>                                   /importer.php
> (JDatabaseEngineMysqlImporter class)
>    libraries/database/engine/postgres/server.php
> (JDatabaseEnginePostgresServer class)
>                                      /query.php
> (JDatabaseEnginePostgresQuery class)
>                                      /exporter.php
> (JDatabaseEnginePostgresExporter class)
>                                      /importer.php
> (JDatabaseEnginePostgresImporter class)
> 
> This will allow the Joomla! autoloader to work without jimport and this
> will be easier for maintaining db engines (one folder for one engine).
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/joomla/joomla-platform/pull/610#issuecomment-3032976

orware committed:
Fixed issue for adding offset/limit in PDO driver (was using empty check instead of is_null check).

orware committed:
Adding SQLite PDO driver.

orware committed:
Adding SQLite Query class.

robschley commented:
I don't like the "engine" approach either as engine is also used to define types of tables such as in MySQL you have MyISAM and InnoDB which are "engines" and we don't want to mix up that terminology.

joomla-jenkins commented:
Build triggered by changes to the head.

Test log missing. Tests failed to execute.
Checkstyle analysis reported 255 warnings and 147 errors.

chdemko commented:
'engine' or 'driver' as you want, my main concern was about files structure

ianmacl commented:
So the question really comes down class names:
JDatabaseDriverMysqlServer vs JDatabaseServerMysql
JDatabaseDriverPostgresQuery vs JDatabaseQueryPostgres

Personally I am in favour of the latter as it means we don't need to have the word driver (or engine) in all of our database classes which could make them more ambiguous as to their actual purpose.

elinw commented:
I'm with Ian on that one. I think it is more readable also.  I'm wondering though about what was raised by Andrew about grouping all the MySQL derived ones together  https://groups.google.com/forum/#!searchin/joomla-dev-platform/drizzle/joomla-dev-platform/cUMgx-351wQ/E4MYobCicdUJ ... I don't know how much of a real issue/convenience that is for any other database but for MySQL we have MySQLi and Drizzle and probably others many of which just have a small number of specific differences but his example of the possibility of different versions is also potentially important. Also Louis had a somewhat different approach earlier in that thread.

orware committed:
Fixed all of the CodeSniffer Errors.

orware commented:
@ianmacl  With the current approach the word "driver" isn't currently in all of the database classes. Some examples:
JDatabaseQueryMySQL
JDatabaseQueryOracle
JDatabaseImporterMySQL
JDatabaseDriverMySQL
JDatabaseDriverOracle

The latest commit should fix all of the CodeSniffer errors...I'm going to take a look at the unit tests next.

orware committed:
Modified class names to conform to the new autoloader standard (e.g. JDatabaseMySQL => JDatabaseMysql, JDatabasePDO => JDatabasePdo)

joomla-jenkins commented:
This pull request could not be tested since the changes could not be cleanly merged.

orware committed:
Changing PHP_EOL constant to simply the newline character so that the unit tests will run correctly on Windows.

orware committed:
Removed unneeded code that's now taken care of by the autoloader. Made sure that the type (driver) is run through ucfirst() (required by the autoloader).

orware committed:
Modified Database Package Unit Tests to work with the new autoloader changes.
Removed unnecessary require_once calls due to the autoloader.

orware committed:
Added in the execute() method into the JDatabaseMock file.

orware committed:
Removed unnecessary require_once calls because of the new autoloader.

orware committed:
Cleaned up a few DocBlock @return comments.
Brought over a bug fix for JTableKeys.

joomla-jenkins commented:
Build triggered by changes to the head.

Unit testing complete.  There were 0 failures and 1 errors from 1882 tests and 11018 assertions.
Checkstyle analysis reported 253 warnings and 0 errors.

orware committed:
Committing change to make the JLoggerDatabase unit test work properly.

orware committed:
Changing instances of query() in the platform to execute().

joomla-jenkins commented:
Build triggered by changes to the head.

Unit testing complete.  There were 0 failures and 0 errors from 1882 tests and 11019 assertions.
Checkstyle analysis reported 253 warnings and 0 errors.

eddieajau commented:
Do we introduce any significant backward compatibility issue with this?

ianmacl commented on libraries/joomla/database/driver/oracle.php:
```diff
@@ -0,0 +1,404 @@
+	 *
+	 * @since   11.4
+	 */
+	public function escape($text, $extra = false)
```
Should this be implemented before we pull to master?

orware commented:
@eddieajau The main things I could think of that would cause possible backwards compatibility issues would be if somebody directly instantiated JDatabaseMySQL classes in the past or something like that (since the autoloader needs the class to be called JDatabaseMysql to instantiate it correctly)...that's mainly because I removed registering the new file locations from all of the files in the database/database directory (I had them initially, but Louis recommended I remove them and just leave the JLog deprecated calls in). Existing MySQL stuff has basically remained untouched, though I have added some new functionality into JDatabaseQuery along with the new Preparable/Limitable interfaces.

@ianmacl I'd definitely love to implement the escape method properly for Oracle, but I simply just don't know exactly what needs to be done for that at the moment, though I could try researching it some more tonight. The main reason why it isn't as big of a concern with the Oracle driver is because prepared statements are used instead so escaping becomes unnecessary. For example:

```
    $query->setQuery('select * from spriden where spriden_pidm = :pidm');
    $query->bind(':pidm', 999999);
    $this->db->setQuery($query);
```

Or even reuse most of your query in a loop and just rebinding the variables:

```
    $insert = range(1, 100);

    $query->setQuery('INSERT INTO tablename (id, requests) VALUES (:idnumber, :requestnumber)');

    foreach($insert as $id)
    {
        $query->bind(':idnumber', $id);
        $query->bind(':requestnumber', $id);
        $this->db->setQuery($query);

        $this->db->execute();
    }
```

Since I think the Oracle class in particular will only be used in custom components with people wanting to tie into another system to retrieve data (mainly), using prepared statements like the ones above will pretty much be required...I don't see us running Joomla on top of Oracle in the near future, but I will try and find some additional information related to escaping and see if I can implement the method more accurately.

orware committed:
Moved the implementation of the escape() method to the Pdo Driver class and removed the basic implementations from the Oracle and Sqlite Driver classes.

orware committed:
Merge branch 'master' of https://github.com/joomla/joomla-platform into merging

joomla-jenkins commented:
Build triggered by changes to the head.

Unit testing complete.  There were 0 failures and 0 errors from 1882 tests and 11019 assertions.
Checkstyle analysis reported 253 warnings and 0 errors.

robschley commented:
The only thing I don't like is that there isn't an actual JDatabaseDriver class. I think if we're going to use the convention of JDatabaseDriverMysql, JDatabaseDriverOracle, JDatabaseDriverSqlsrv, etc. we should rename JDatabase to JDatabaseDriver.

elinw commented:
It seems like we're adding an extra layer of hierarchy for  no reason that has been explained.  is there something else besides a folder named driver that would go into the actual database folder? I think the way it is now shows the problems with not enough hierarchy, but if there is not something else anticipated it doesn't make sense for me.

The bigger question is whether it makes sense to organize by database or object type. I think tifhere is something to be said for having packages by database if the idea is that eventually people will pick the packages they need and leave the rest.

orware commented:
@robschley I was thinking along changing the name initially, but thought against renaming JDatabase for two reasons:
1. Mainly, I liked JDatabase over JDatabaseDriver
2. JDatabase feels more like a factory in a way for getting instances of any driver, query object, importer, exporter, etc. so that made me feel like it wasn't necessary to add the Driver bit at the end.

@elinw Here's a comparison between what we have currently:

```
joomla/
    database/
        database/
        table/
```

So in the current scenario we have driver classes, query classes, and the more recent importer/exporter classes all residing in the database/database folder.

In the new organization we have the following:

```
joomla/
    database/
        database/ (for the backwards compatible classes)
        driver/
        exporter/
        importer/
        query/
        table/
```

For me I don't feel that's much of an extra hierarchy, and I enjoy the fact that now all of the files are organized in a way that makes sense, at least for me. And actually, most of this reorganization is really related to making the needed changes so that the autoloader would work properly, and in a way that didn't modify much of the class names (aside from the driver ones, I didn't have to rename the query, exporter or importer classes...aside from the casing change I had to make the other day from "JDatabaseMySQL" to "JDatabaseMysql" since the autoloader uses camel casing as part of its convention for locating files.

elinw commented:
Yeah I see how you are thinking and that structure doesn't have the extra level. Happy to lose the database/database.

orware committed:
Merge branch 'master' of https://github.com/joomla/joomla-platform into merging
Added in Andrew's reverting of Pull Request #570

Conflicts:
	libraries/joomla/database/database/mysqli.php

joomla-jenkins commented:
Build triggered by changes to the head.

Unit testing complete.  There were 0 failures and 0 errors from 1882 tests and 11019 assertions.
Checkstyle analysis reported 253 warnings and 0 errors.

orware committed:
Made the extra part in JDatabaseDriverPdo::escape() always be included.

joomla-jenkins commented:
Build triggered by changes to the head.

Unit testing complete.  There were 0 failures and 0 errors from 1882 tests and 11019 assertions.
Checkstyle analysis reported 253 warnings and 0 errors.

orware committed:
Adding the new JDatabaseFactory class.
Adding JDatabaseDriver class (replacing JDatabase)
Adding JDatabaseExporter
Adding JDatabaseImporter

orware committed:
Converted references to JDatabase to JDatabaseDriver in comments.

orware committed:
Changed JDatabase to extend from JDatabaseDriver and deprecated it for 11.4.

orware committed:
Moved the connect logic out of the driver constructors and into a new connect() method.
Now any JDatabaseDriver methods that talk to the database should call connect() first before running any queries to ensure that a connection has been created.

orware committed:
Fixed any classname references that were still incorrect (changed to new camel casing form necessary for autoloader).
Made the Importer/Exporter classes extend from JDatabaseImporter/JDatabaseExporter, respectively.

orware committed:
Added str_replace to the testAddElementPath() method so the test would work on Windows.

orware committed:
Added the new connect method for the JDatabaseMock.

orware committed:
Fixed unit tests by making sure the correct class names are being used.

joomla-jenkins commented:
Build triggered by changes to the head.

Unit testing complete.  There were 1 failures and 0 errors from 1882 tests and 11019 assertions.
Checkstyle analysis reported 253 warnings and 2 errors.

orware committed:
Fixed two CodeSniffer errors.

orware committed:
Removed the str_replace() added earlier today to see if the if it will fix running the unit tests from Jools.

joomla-jenkins commented:
Build triggered by changes to the head.

Unit testing complete.  There were 0 failures and 0 errors from 1882 tests and 11019 assertions.
Checkstyle analysis reported 253 warnings and 0 errors.

eddieajau commented:
Note added to title (remove on merge). We won't consider this until after 11.4.

gpongelli commented:
@orware 
Why don't you use "git mv" instead of moving files directly?
With former command you can keep history of moved files, that I think it's better.
Look at [my commit](https://github.com/joomla/joomla-platform/pull/532/files#diff-10), it is highlighted as
<b>tests/{ddl.sql → ddl_mysql.sql}</b>
instead of simply removing one file and adding the same with another path or name.
Hoping to see this changes in another pull request ;)

Bye,
Eng. Gabriele Pongelli

gpongelli commented:
Just to be clear, "git mv" allow other developer (maintainer of master branch too) to do a rebase without losing their changes on moved files instead of merging and have conflicts/painful state.
I've tried to merge this branch on a copy of my postgresql branch, the result was a conflict state.
Then I've tried rebasing these commits on another copy of my branch to keep my changes and add these one after mine, the result was conflict state worse than previous.

Using "git mv" instead, with rebasing --onto and then merge (as explained [here](http://learn.github.com/p/rebasing.html) ), the result is a fast-forward commit, without any conflict.

These are the commands I've done:

<pre><code># git branch base_br master
# git branch moving master
# git checkout moving
Switched to branch 'moving'
# cd libraries/joomla/database/database
# git mv mysql.php ../
# cd ../../
# git commit -m "moved file"
[moving 79f0b06] moved file
 1 files changed, 0 insertions(+), 0 deletions(-)
 rename libraries/joomla/database/{ => }/database/mysql.php (100%)
# git rebase --onto base_br base_br moving
Already on 'moving'
Current branch moving is up to date.
# git checkout base_br
Switched to branch 'base_br'
# git merge moving
Updating bdfe082..79f0b06
Fast-forward
 .../joomla/database/{ => }/database/mysql.php      |    0
 1 files changed, 0 insertions(+), 0 deletions(-)
 rename libraries/joomla/database/{ => }/database/mysql.php (100%)</code></pre>


As you can see there's no conflicts, it's a clear solution and it will not create problems to other developer's workspace ;)

Eng. Gabriele Pongelli.

elinw commented:
Currently this pull request is out of sync with a number of impacted files. 
I would like to make the following proposal.
This should be broken into two  phases or steps.
Step one should be the introduction of the new code in the present location, database/database.
Once those are merged do Step Two which would be to move all of the files to the new locations without making anadditional changes in the same action as moving.
Merge those.
Then we will not have to worry about regressions due to having to hand merge all f the changes from the current libraries not to mention the loss of history. Done carefully I think it can be seamless. Personally, if the code is good I have no way of testing) I have no problem at all with having additional classes and methods available in 11.4, but  I think rearranging thngs probably rightly should be done after 11.4. It just adds a level of complexity that is unnecessary in the short term.

gpongelli commented:
I agree on Elin thought, hoping that moving action will be done with "git mv" ;) .

chdemko commented:
Fixed in #906
